tools/testbench/common_test_ipc4.c

```diff
@@ -0,0 +1,593 @@
+
+#if CONFIG_IPC_MAJOR_4
+
+#include <sof/audio/component_ext.h>
+#include <sof/lib/notifier.h>
+#include <sof/audio/component_ext.h>
+#include <sof/schedule/edf_schedule.h>
+#include <sof/schedule/ll_schedule.h>
+#include <sof/schedule/ll_schedule_domain.h>
+#include <ipc/stream.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "testbench/common_test.h"
+#include "testbench/file.h"
+#include "testbench/topology_ipc4.h"
+
+#if defined __XCC__
+#include <xtensa/tie/xt_timer.h>
+#endif
+
+SOF_DEFINE_REG_UUID(testbench);
+DECLARE_TR_CTX(testbench_tr, SOF_UUID(testbench_uuid), LOG_LEVEL_INFO);
+LOG_MODULE_REGISTER(testbench, CONFIG_SOF_LOG_LEVEL);
+
+/* testbench helper functions for pipeline setup and trigger */
+
+int tb_setup(struct sof *sof, struct testbench_prm *tp)
+{
+	struct ll_schedule_domain domain = {0};
+	int bits;
+	int krate;
+
+	domain.next_tick = tp->tick_period_us;
+
+	/* init components */
+	sys_comp_init(sof);
+
+	/* Module adapter components */
+	sys_comp_module_crossover_interface_init();
+	sys_comp_module_dcblock_interface_init();
+	sys_comp_module_demux_interface_init();
+	sys_comp_module_drc_interface_init();
+	sys_comp_module_eq_fir_interface_init();
+	sys_comp_module_eq_iir_interface_init();
+	sys_comp_module_file_interface_init();
+	sys_comp_module_gain_interface_init();
+	sys_comp_module_google_rtc_audio_processing_interface_init();
+	sys_comp_module_igo_nr_interface_init();
+	sys_comp_module_mfcc_interface_init();
+	sys_comp_module_multiband_drc_interface_init();
+	sys_comp_module_mux_interface_init();
+	sys_comp_module_rtnr_interface_init();
+	sys_comp_module_selector_interface_init();
+	sys_comp_module_src_interface_init();
+	sys_comp_module_asrc_interface_init();
+	sys_comp_module_tdfb_interface_init();
+	sys_comp_module_volume_interface_init();
+
+	/* other necessary initializations, todo: follow better SOF init */
+	pipeline_posn_init(sof);
+	init_system_notify(sof);
+
+	/* init IPC */
+	if (ipc_init(sof) < 0) {
+		fprintf(stderr, "error: IPC init\n");
+		return -EINVAL;
+	}
+
+	/* Trace */
+	ipc_tr.level = LOG_LEVEL_INFO;
+	ipc_tr.uuid_p = SOF_UUID(testbench_uuid);
+
+	/* init LL scheduler */
+	if (scheduler_init_ll(&domain) < 0) {
+		fprintf(stderr, "error: edf scheduler init\n");
+		return -EINVAL;
+	}
+
+	/* init EDF scheduler */
+	if (scheduler_init_edf() < 0) {
+		fprintf(stderr, "error: edf scheduler init\n");
+		return -EINVAL;
+	}
+
+	tb_debug_print("ipc and scheduler initialized\n");
+
+	/* setup IPC4 audio format */
+	tp->num_configs = 1;
+	krate = tp->fs_in / 1000;
+	switch (tp->frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		bits = 16;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		bits = 24;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		bits = 32;
+		break;
+	default:
+		fprintf(stderr, "error: unsupported frame format %d\n", tp->frame_fmt);
+		return -EINVAL;
+	}
+
+	/* TODO 44.1 kHz like rates */
```

Comment:
make a bug for this, orphaned todos in code are useless

Reply:
Ah, that's something I have not yet been able to test, 44.1 kHz capability is very useful so it won't be orphaned. Also seems that copier in/out formats parsing is not complete, only 1st format is working (S32_LE) so even basic S16_LE fails with the topologies I used.

Reply:
This and next one remain still opens in the version of today. I will address them later when I find suitable topologies to use and have found fix for copier formats discovery.